lxtools/lxbackup.1
.TH lxbackup l Local
.SH NAME
lxbackup - Backup and restore the files on your HP 95LX, 100LX or 200LX
.SH SYNOPSIS
.B lxbackup
[
.I dir ...
] [
.B -all
] [
.B -since
.I mmddhhmm[yy]
] [
.B -newer
.I filename
] [
.B -a
] [
.B -v
] [
.B -n
] [
.B -c
] [
.I -port
] [
.B -d
.I devname
] [
.B -b
.I baudrate
] [
.B -l
] [
.B -restore
] [
.B -q|h|?
] [
.B -exclude
.I pat pat ...
]

.SH DESCRIPTION
.I lxbackup
is a program to automatically do backups and restores on the
HP100LX and HP200LX palmtop computers using a serial line. In order to use
lxbackup, you need to connect your palmtop with the Unix machine using a
serial cable. Then, you have to launch Filer on the palmtop and adjust the
communication options in the "Communications/Remote Settings..." dialog
of Filer to 38400 baud, Com1 and server mode enabled. That's it.

The default is to backup any file on the HP which is newer than its
most recent backup copy (if any).  Optionally, you may backup all files
newer than a specified date, newer than a specified file, or all
files regardless of age.

.SH OPTIONS
.TP
.B -q|h|?
Gets quick help.
.TP
.B -all
Tells lxbackup to backup all files, regardless of age.
.TP
.B -a
Tells lxbackup to backup all files with archive bit set, ignoring
others.  The archive bit is set whenever you modify a file. It can
be cleared using  the  ATTRIB command (see ATTRIB /?).
.TP
\fB-since\fP \fImmddhhmm[yy]\fP
Tells lxbackup to backup all files modified more recently than the
specified date. (yy < 80 is assumed to be 21st century, e.g. 
"1231094500" would specify 9:45 am, Dec 31st, 2000
.TP
\fB-newer\fP \fIfilename\fP
Tells lxbackup to backup all files modified more recently than the
specified file.
.TP
.B -v
Tells lxbackup to output status messages as it performs backups.
.TP
.B -n
Tells lxbackup not to actually transfer any files, but to list the
files that \fIwould\fP be transferred otherwise.
.TP
.B -c
Tells lxbackup to use "-" instead of ":" when writing files to the
local filesystem. Useful when writing to DOS partitions, NetWare
servers, etc. Note to restore from such a backup you must also use
the -c switch, like this:
.nf

	lxbackup -c -restore c:

.TP
.B -restore
Tells lxbackup to transfer files from the Unix system to the HP LX.
.TP
.B -1, -2, -3 ... -8
Sets serial line 1, 2, 3 ... 8, respectively. Default is 1.
These correspond to /dev/cua0, /dev/cua1 and so on on some computers,
/dev/ttya, /dev/ttyb and so on on others.
.TP
\fB-d /dev/\fP\fIdevice\fP
Sets serial line explicitly.
.TP
\fB-b\fP \fIbaudrate\fP
sets the baud rate to \fIbaudrate\fP. Default is 38400.
.TP
\fB-exclude\fP \fIpattern pattern ...\fP
Provides a list of patterns to exclude from backup (or restore).  Files
matching these patterns will not be transferred.  Typically used to
exclude .exe, .exm and .com files, which are usually large and which are
probably backed up somewhere else already.



.SH EXAMPLES
.TP
lxbackup
Backup your C: disk, baud rate defaults to 38400, port defaults to
first port, directory defaults to C:, backup everything in sight.
.TP
lxbackup c: -2 -baud 19200
Backup your C: disk, use port 2, set baud rate to 19200.
.TP
lxbackup c: a: -n -exclude .exe .exm .com
Check your C: and A: disks, see which files would be backed up.
Skip any files ending with .exe, .exm, or .com
.TP
lxbackup c: a: -v -exclude .exe .exm .com
Backup your C: and A: disks.  Watch progress of backup.
Skip any files ending with .exe, .exm, or .com
.TP
lxbackup c:/_dat c:/games c:/pgp -v -exclude .exe .com
Backup some specific directories.

.TP
Shell Script
Here is the shell script you can use on your unix box to execute your
daily backups.
.nf

	#! /bin/sh

	cd $HOME/hp100/backups
	lxbackup c: a: -v -d /dev/ttyb -exclude .exm .exe .com

.fi
(This creates the subdirectories $HOME/hp100/backups/c:/ and
$HOME/hp100/backups/a:/)

.TP
lxbackup -restore
Restore everything from your C: disk.  Baud rate defaults to 38400, port
defaults to first port.
.TP
lxbackup c:/games -2 -baud 19200
Restore one specific directory.
.TP
lxbackup c: -v -restore -newer lastbackup.list
Restore all files which have been modified on local disk more recently
than the file "lastbackup.list".

.SH NOTES

Upper case and lower case do not matter.  Lxbackup will convert
all filenames to upper case for the HP, and lower case for the
local file systems.
.PP
Likewise, it does not matter if you use '/' or '\' as the
directory seperator.
.PP
It is not required to put the drive letter in the directory
name, but highly recommended.  Whichever you do, be consistent,
as this is the name lxbackup will use for the directory into
which the files will be stored.  If you use a drive letter, the
backups will be done into a directory with that drive letter as
a name, e.g. "c:/".  (Unix allows ':' (and practically anything
else) as part of a directory or file name)
.PP
Order of directories and options is unimportant, except that
the -exclude option must be the last thing on the command
line.
.PP   
If you have problems with transfering files, you can try a lower
baud rate. Set your palmtop, e.g., to 19200 baud and run the
LX tools with the "-b 19200" option.

.SH ENVIRONMENT

The device name and number as well as the baud rate can be overriden
by environment variables. If these variables are set, the corresponding
command line options are ignored:
.TP
LXTOOLS_LINE
specifies the device to be used, e.g. "/dev/ttya".
.TP
LXTOOLS_BAUD
specifies the baud rate to be used, e.g. "9600".
.TP
LXTOOLS_STDOUT
specifies that output should be sent to STDOUT (default is STDERR).

.SH DIAGNOSTICS

If you see the message "Cannot open /dev/cua0", most probably you have no

lxfileman/Changelog

Thanks go out to Mitch Hamm and Jaime Viehweg for helping debug!

Change history of LXFileman:

0.96b  8/01/1999 Fixed:
		 * Previous versions would crash if they tried to list
		   the contents of a Unix directory that contained a
		   file with curly braces (as created by 'find2perl')

0.96a  3/06/1999 Fixed:
		 * another stupid port/speed problem when double-clicking
		   on file listings (yes, Mitch, finally)

0.96   2/21/1999 Fixed:
		 * Problems setting port/speed when using LX
		   drive letter buttons
		 Added:
		 * Checks OS name with uname and handles Linux
		   'setserial' port speed configuration

0.91c 11/24/1997 Fixed:
		 * Stray diagnostic message when Delete key pressed
		 Added:
		 * Much nicer dialog boxes for deletion confirmation

0.91b 11/9/1997	 Fixed:
		 * F2 & Delete key bindings really messed up
		   dialog boxes, also interfered with regular
		   LX & UNIX location entry boxes
		 * Dump if unable to "cd" into requested UNIX dir
		 * Did not restore LX location if LX listing failed

0.91  11/9/1997  Fixed:
		 * Cleaned the code up a bit, trimmed script
		   filesize by ~ 10%
		 Added:
		 * Custom port name now shows up in cascading menu
		 * Cancel button on Set custom port name dialog
		 * Changed many radiobuttons to checkbuttons,
		   added more color, other general UI improvements

0.90c 11/6/1997  Added:
		 * couple more bindings that change the cursor

0.90b 11/2/1997  Added:
		 * Now changes cursor to watch icon when doing most
		   time-consuming operations
		 * Now has bindings for F2 (copy) and Delete keys
		 * Now has optional delete confirmation

0.90 10/12/1997  Fixed (?)
		 * Change Button binds to "ButtonRelease-1" per 
		   suggestion of a USENET post by a Sun employee
		   ...should fix Copy/Delete button corruption

0.89 8/23/1997	 Fixed:
	 	 * Now parses output of ls on startup so more versions
		   of "ls" should be auto-detected
		 * Variables & functions renamed for better readability

0.88e 8/10/1997  Fixed:
		 * LX display broken if there was a filename with a space
		   (note that LXCOPY and even the Filer have trouble with
		    filenames that include spaces)
		 Added: 
		 * Option to sort LX files & dir names alphabetically
		 * "Cancel" button on New Dir dialogue box
		 * Cleaned up the Settings menus some more

0.88c 8/7/1997	 Added:
		 * Colors on dialog bitmaps
		 * Settings saves size & window postion
		 * Added UNIX sorting option
		 * Made 4 UNIX options a cascading menu

0.88 8/5/1997	 Fixed:
		 * 0.87's ls -laF code did not work with file/dir names
		   that had spaces 
		 * Dialog boxes now appear in predictable places
		 * Main window now resizes well horizontally

0.87 8/4/1997	 Fixed:
		 * did not get name from ls -laF correctly on Solaris
		 Added:
		 * Drive button letters now user-configurable

lxtools/lxbackup.c
/* $Id: lxbackup.c,v 1.2 2000/10/09 00:08:02 gonter Exp $ */
/* --------------------------------------------------------------------
	Project: Communication package Linux-HPx00LX Filer
	Author:  Edward A. Falk
	Started: 12 Mar 97
	Subject: HP95/HP100/HP200 incremental backup utility
  -------------------------------------------------------------------- */

/* -------------------------------------------------------------------
	Modifications by Peter Watkins
	November 16, 1997
	- added support for LXTOOLS_STDOUT
	- added "-c" switch (and added notes to man page)
	- removed hard-coded exclude rule
	- corrected notReally error in restore code
   ------------------------------------------------------------------- */

/* -------------------------------------------------------------------
	Modifications by Peter Watkins
	January 1, 2000
	- changed date functions to assume that -since dates < 80
	  or DOS timestamps with years < 80 are 21st century
   ------------------------------------------------------------------- */


static	char	*usage =

"Backup files on hp95, hp100 or hp200, via Filer\n"
"Usage: lxbackup [options] [dir...]\n"
"\n"
"	<dir> <dir>...		disks & directories to back up (C:\\)\n"
"\n"
"	-all			all files.\n"
"	-since <mmddhhmm[yy]>	all files modified since date\n"
"	-newer <filename>	all files newer than specified file\n"
"	-a			all files that have the archive bit set\n"
"	-1|2|3|4		set com port (1)\n"
"	-d <devname>		set full com port name\n"
"	-b <baud>		set baud rate (38400)\n"
"	-l			list remote files only\n"
"	-n			Show what would be done only.\n"
"	-c			Use \"-\" instead of \":\" for local files\n"
"	-v			verbose\n"
"	-restore		restore instead of backup\n"
"	-q|h|?			this message\n"
"	-exclude <pat...>	exclude files matching patterns (must be last)\n"
"\n"
" Return values:\n"
"	0 = success\n"
"	1 = failure\n"
"	2 = user error\n"
"	3 = internal error\n"
	;

#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <time.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>
#include <ftw.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/param.h>
#include <sys/stat.h>

#include "pal.h"
#include "config.h"

#define	NAMESIZE	13		/* Dos filenames limited to 12 chars */
#define	MAXDIRS		256		/* arbitrary. */

extern	int	sys_nerr ;
/* commented out for LXTools 1.1c
extern	char	*sys_errlist[] ;
*/

static	char	*excludes[1024] ;
static	int	nexcludes = 0 ;

static	time_t	since, parseDate(), fileTime() ;
static	int	archive = 0 ;
static	int	verbose = 0 ;
static	int	port = 1 ;
static	char	*dev_name = NULL ;
static	int	baud = 38400 ;
static	int	listOnly = 0 ;
static	char	*top[MAXDIRS] ;
static	int	ntop = 0 ;
static	int	compress = 0 ;
static	int	restore = 0 ;
static	int	notReally = 0 ;
static	int	noColon = 0 ;


typedef	struct stackedName {
	  char	name[NAMESIZE] ;
	  FDATESTAMP date ;
	  unsigned long size ;
	  struct stackedName	*next ;
	} StackedName ;

#define	strmatch(a,b)	(strcmp(a,b)==0)
#define	MALLOC(t,n)	((t *)malloc(sizeof(t)*(n)))

/* commented out for LXTools 1.1c
extern	int	putenv(char *) ;
*/

	/* forward references */


static	int	BackupDir(FILERCOM *, char *dirName) ;
static	int	RestoreDir(FILERCOM *, char *dirName) ;
static	int	RestoreCB(const char *, const struct stat *, int) ;
static	int	RestoreFile(FILERCOM *, const char *, const struct stat *) ;
static	int	CreateDir(FILERCOM *pFiler, const char *name) ;
static	int	makedir(char *) ;
static	int	CheckFile(FILERCOM *pFiler, char *name) ;
static	int	ExcludeFile(const char *name) ;
static	int	BackupFile(FILERCOM *pFiler, char *name, StackedName *) ;
static	int	makePath(char *name) ;
static	time_t	FDate2Unix(FDATESTAMP *) ;
static	void	dos2unix(const char *dos, char *unx) ;
static	void	unix2dos(const char *unx, char *dos) ;
static	const char	*lxbasename(const char *) ;
static	void	die(char *msg, char *msg2, int rcode) ;
static	void	colonToDash(const char *unx, char *dos) ;


int
main(int argc, char **argv)
{
	int	i,j ;
	char	topDir[MAXPATHLEN] ;
	char	ttyEnvBuf[256] ;
	FILERCOM *pFiler ;
	int	rval = 0 ;
	char	goodName[MAXPATHLEN] ;

	/* couldn't see that it respected LXTOOLS_BAUD */	
	if(getenv("LXTOOLS_BAUD")) {
	  if(atoi(getenv("LXTOOLS_BAUD"))) {
	    baud = atoi(getenv("LXTOOLS_BAUD"));
	  } 
	}

	since = -1 ;

	while(--argc > 0)
	{
	  ++argv ;
	  if( **argv == '-' )
	  {
	    if( strmatch(*argv, "-since") ) {	/* mmddhhmm[yy] format */
	      if( --argc > 0 )
		since = parseDate(*++argv) ;
	      else
		die("-since requires a date",NULL,2) ;
	    }
	    else if( strmatch(*argv, "-newer") ) {
	      if( --argc > 0 )
		since = fileTime(*++argv) ;
	      else
		die("-newer requires a file name",NULL,2) ;
	    }
	    else if( strmatch(*argv, "-all") )
	      since = 0 ;
	    else if( strmatch(*argv, "-a") )
	      archive = 1 ;
	    else if( strmatch(*argv, "-v") )
	      ++verbose ;
	    else if( strmatch(*argv, "-restore") )
	      restore = 1 ;
	    else if( strmatch(*argv, "-z") )
	      compress = 1 ;
	    else if( strmatch(*argv, "-l") )
	      listOnly = 1 ;
	    else if( strmatch(*argv, "-n") )
	      verbose = notReally = 1 ;
	    else if( strmatch(*argv, "-c") )
	      noColon = 1 ;
	    else if( isdigit(argv[0][1]) )
	      port = atoi(&argv[0][1]) ;
	    else if( strmatch(*argv, "-d") && --argc > 0 ) {
	      dev_name = *++argv ;
	      sprintf(ttyEnvBuf, "LXTOOLS_LINE=%s", dev_name) ;
	      putenv(ttyEnvBuf) ;
	    }
	    else if( strmatch(*argv, "-b") && --argc > 0 )
	      baud = atoi(*++argv) ;

	    else if( strmatch(*argv, "-exclude") )
	    {
	      --argc ; ++argv ;
	      while( argc > 0 && **argv != '-' )
	      {
		char *ptr ;
		excludes[nexcludes++] = ptr = *argv++ ;
		for( ;*ptr != '\0'; ++ptr )
		  *ptr = toupper(*ptr) ;
		--argc ;
	      }
	    }

	    else if(strmatch(*argv,"-?") || strmatch(*argv,"-h") ||
		    strmatch(*argv,"-q") )
	      die(usage,NULL,0) ;

	    else {
	      fprintf(USE_OUT, "unknown option: %s\n",*argv) ;
	      die(usage,NULL,2) ;
	    }

	  }
	  else if( ntop < MAXDIRS )
	    top[ntop++] = *argv ;
	  else {
	    fprintf(USE_OUT,
	      "lxbackup: too many directories specified (max %d), %s ignored\n",
	      MAXDIRS, *argv) ;
	    rval = 3 ;
	  }
	}

	if( verbose >= 2 ) fprintf(USE_OUT,"connecting to remote...") ;
	if( !(pFiler = FilerConnect(port, baud, &FlCb)) )
	  die("failed to connect to remote",NULL,1) ;
	if( verbose >= 2 ) fprintf(USE_OUT,"connected\n") ;

	if( ntop == 0 )
	  top[ntop++] = "C:\\" ;


	for(i=0; i<ntop; ++i)
	{
	  if( !restore )
	  {
	    strncpy(topDir, top[i], sizeof(topDir)) ;
	    makedir(topDir) ;
	    if( (j = BackupDir(pFiler, topDir)) )
	      rval = j ;
	  }
	  else
	  {
	    colonToDash(top[i],goodName);
/*
	    fprintf(USE_OUT,"passing \"%s\" to RestoreDir\n",goodName);
*/
	    if( (j = RestoreDir(pFiler, goodName)) )
	      rval = j ;
	  }
	}


	if( verbose >= 2 ) fprintf(USE_OUT,"disconnecting ...") ;
	FilerDisconnect(pFiler) ;
	if( verbose ) fprintf(USE_OUT, "done\n") ;

	exit(EXITVAL(rval)) ;
}



	/* convert a directory name to "canonical" form.  That is,
	 * all '/' converted to '\', and a trailing '\' added if needed.
	 */

static	int
makedir(char *dir)
{
	char *p ;

	for(p = dir; *p; ++p )
	  if( *p == '/' ) *p = '\\' ;
	if( p[-1] != '\\' ) *p++ = '\\' ;
	return p - dir ;
}



	/* This is the heart of the backup procedure.  Given a directory
	 * name, ask for a listing of that directory.  All files are
	 * tested with CheckFile(), all directories are stacked up.
	 *
	 * Once the listing is complete, all flagged files passed to
	 * BackupFile(), all directories are passed (recursively) to
	 * BackupDir()
	 *
	 * Input name must always have three extra bytes at the end so
	 * that this function may append "*.*".  Input name must be in
	 * canonical form, described in makedir().
	 */

static	int
BackupDir(FILERCOM *pFiler, char *name)
{
	StackedName	*dirStack = NULL ;
	StackedName	*fileStack = NULL ;
	StackedName	*namePtr ;
	int		len = strlen(name) ;
	int		status ;
	int		rval = 0 ;

	if( len > MAXPATHLEN-5 ) {
	  fprintf(USE_OUT, "Filename too long: %s\n", name) ;
	  return 3 ;
	}

	strcpy(name+len, "*.*") ;
	status = FilerAskDir(pFiler, name);
	name[len] = '\0' ;

	switch( status ) {
	 case NO_RESPONSE:
	    fprintf(USE_OUT, "Server not responding.\n") ;
	    return 1 ;
	 case BAD_REQUEST:
	    fprintf(USE_OUT, "Bad request.\n") ;
	    return 1 ;
	 default:
	    fprintf(USE_OUT, "Transmission error.\n") ;
	    return 1 ;
	 case SERVER_ACK:
	    break ;
	}

	if( listOnly )
	  printf("\nDirectory of %s:\n\n", name) ;
	else if( verbose >= 2 )
	  fprintf(USE_OUT, "Searching directory %s:\n", name) ;

	while( FilerGetDir(pFiler) != CANNOT_GET_ENTRY )
	{
	  if( pFiler->Attribute & 0x10 )
	  {
	    /* It's a directory.  Save the name for later. */
	    if( strcmp((char *)pFiler->Name, ".") != 0  &&
	        strcmp((char *)pFiler->Name, "..") != 0 )
	    {
	      namePtr = MALLOC(StackedName, 1) ;
	      if( namePtr == NULL ) {
		fprintf(USE_OUT, "lxbackup: out of memory\n") ;
		return 3 ;
	      }

	      strcpy(namePtr->name, (char *)pFiler->Name) ;
	      namePtr->next = dirStack ;
	      dirStack = namePtr ;
	    }

	    if( listOnly )
	      printf("%-12s    <DIR>      %02d-%02d-%02d  %02d:%02d\n",
		pFiler->Name,
		pFiler->DateStamp.month, pFiler->DateStamp.day,
		pFiler->DateStamp.year+80,
		pFiler->DateStamp.hour, pFiler->DateStamp.min);
	  }
	  else
	  {
	    /* It's a file.  See if it needs to be backed up. */
	    if( listOnly )
	      printf("%-12s %12lu  %02d-%02d-%02d  %02d:%02d\n",
		pFiler->Name, pFiler->FileSize,
		pFiler->DateStamp.month, pFiler->DateStamp.day,
		pFiler->DateStamp.year+80,
		pFiler->DateStamp.hour, pFiler->DateStamp.min);
	    else
	    {
	      if( len+strlen((char *)pFiler->Name) >= MAXPATHLEN )
		fprintf(USE_OUT, "File name too long (ignored): %s\\%s\n",
		    name, pFiler->Name) ;
	      else
	      {
		strcpy(name+len, (char *)pFiler->Name) ;
		if( CheckFile(pFiler, name) )
		{
		  namePtr = MALLOC(StackedName, 1) ;
		  if( namePtr == NULL ) {
		    fprintf(USE_OUT, "lxbackup: out of memory\n") ;
		    return 3 ;
		  }

		  strcpy(namePtr->name, (char *)pFiler->Name) ;
		  namePtr->date = pFiler->DateStamp ;
		  namePtr->size = pFiler->FileSize ;
		  namePtr->next = fileStack ;
		  fileStack = namePtr ;
		}
	      }
	    }
	  }
	}


	/* Next: unstack the files and back them up. */

	while( rval == 0 && fileStack != NULL )
	{
	  namePtr = fileStack ;
	  fileStack = fileStack->next ;

	  strcpy(name+len, namePtr->name) ;
	  rval = BackupFile(pFiler, name, namePtr) ;
	  name[len] = '\0' ;

	  free(namePtr) ;
	}


	/* Next: unstack the directories and recurse */

	while( rval == 0 && dirStack != NULL )
	{
	  namePtr = dirStack ;
	  dirStack = dirStack->next ;
	  if( len+strlen(namePtr->name) >= MAXPATHLEN )
	    fprintf(USE_OUT, "Directory name too long (ignored): %s\\%s\n",
		name, namePtr->name) ;
	  else
	  {
	    strcpy(name+len, namePtr->name) ;
	    strcat(name+len, "\\") ;
	    rval = BackupDir(pFiler, name) ;
	    name[len] = '\0' ;
	  }

	  free(namePtr) ;
	}

	return rval ;
}



	/* examine filename and determine if it needs to be backed up */

static	int
CheckFile(FILERCOM *pFiler, char *name)
{
	char	name2[MAXPATHLEN] ;
	time_t	ftime ;

	if( ExcludeFile(name) )
	  return 0 ;

	if( archive )
	  return pFiler->Attribute & 0x20 ;

	dos2unix(name, name2) ;

	if( since != 0 )	/* convert filer date to unix date */
	{
	  ftime = FDate2Unix(&pFiler->DateStamp) ;

	  if( since == -1 )	/* compare file dates */
	  {
	    if( ftime <= fileTime(name2) )
	      return 0 ;
	  }
	  else
	  {
	    if( ftime <= since )
	      return 0 ;
	  }
	}


	if( !notReally && makePath(name2) )
	  return 0 ;

	return 1 ;
}



static	int
ExcludeFile(const char *name)
{
	char	**ptr = excludes ;
	int	len = strlen(name) ;
	int	n = nexcludes ;

/* hard-coded: exclude directories */
	if( strncmp(name+len-1,"\\",4) == 0 )
	  return 1 ;

/* hard-coded: exclude directories, c:\_*.wk1 
	if( strncmp(name+len-1,"\\",4) == 0 ||
	    strncmp(name,"C:\\_",4) == 0 && strncmp(name+len-4,".WK1",4) == 0 )
	  return 1 ;
*/

	/* search exclude list */

	for(; --n >= 0; ++ptr )
	  if( strstr(name, *ptr) != NULL )
	    return 1 ;

	return 0 ;
}



static	int
BackupFile(FILERCOM *pFiler, char *name, StackedName *namePtr)
{
	char	name2[MAXPATHLEN] ;
	struct timeval times[2] ;

	dos2unix(name, name2) ;
	if( verbose )
	  fprintf(USE_OUT, "backup %s, %ld bytes ... ", name, namePtr->size) ;

	if( notReally )
	  fprintf(USE_OUT, "\n") ;
	else
	{
	  if( FilerGetFile(pFiler, name, name2) != GOT_FILE_OK )
	    return 1 ;

	  times[0].tv_sec = times[1].tv_sec = FDate2Unix(&namePtr->date) ;
	  times[0].tv_usec = times[1].tv_usec = 0 ;
	  (void) utimes(name2, times) ;

	  if( compress )
	  {
	    char cmd[MAXPATHLEN] ;
	    strcpy(cmd, "gzip ") ;
	    strncat(cmd, name2, sizeof(cmd)) ;
	    if( verbose ) fprintf(USE_OUT, "compress ... ") ;
	    system(cmd) ;
	  }
	  if( verbose ) fprintf(USE_OUT, "done\n") ;
	}
	return 0 ;
}



	/* This is the heart of the restore procedure.  Given a directory
	 * name, examine that directory and its contents.  For each file
	 * found, see if it exists or is older on the HP.  If so,
	 * write it to the HP.
	 */

static	FILERCOM	*fcom ;	/* stupid ftw() doesn't accept client args */

static	int
RestoreDir(FILERCOM *pFiler, char *name)
{
	fcom = pFiler ;
	ftw(name, RestoreCB, 10) ;
	return 0 ;
}


static	int
RestoreCB(const char *name, const struct stat *sbuf, int what)
{
/*
	fprintf(USE_OUT,"RestoreCB was passed \"%s\"\n",name);
*/
	switch( what ) {
	  case FTW_F:
#ifdef	FTW_SL
	  case FTW_SL:
#endif
	    return RestoreFile(fcom, name, sbuf) ;

	  case FTW_D:
#ifdef	FTW_DP
	  case FTW_DP:
#endif
	    if( verbose >= 2 )
	      fprintf(USE_OUT, "Directory %s:\n", name) ;
	    CreateDir(fcom, name) ;
	    break ;

	  case FTW_DNR:
	    fprintf(USE_OUT, "directory %s unreadable\n", name) ;
	    break ;
	  case FTW_NS:
#ifdef	FTW_SLN
	  case FTW_SLN:
#endif
	    fprintf(USE_OUT, "file %s unreadable\n", name) ;
	    break ;
	}
	return 0 ;
}



	/* conditionally create a directory on the HP */

static	int
CreateDir(FILERCOM *pFiler, const char *name)
{
	char	dosname[MAXPATHLEN] ;
	const char	*dirname ;
	int	needCreate = 1 ;

	if( strlen(name) == 2 && name[1] == ':' )
	  return 0 ;

	unix2dos(name, dosname) ;
	dirname = lxbasename(dosname) ;

	if( FilerAskDir(pFiler, dosname) != SERVER_ACK )
	  return 1 ;

	while( FilerGetDir(pFiler) == GOT_DIR_ENTRY )
	  if( strmatch((char *)pFiler->Name, dirname) )
	  {
	    if( pFiler->Attribute & 0x10 )
	      needCreate = 0 ;
	    else
	      fprintf(USE_OUT, "Cannot create directory %s on top of file\n",
		dosname) ;
	  }

	if( needCreate )
	{
	  if( verbose ) fprintf(USE_OUT, "Create directory %s\n", dosname) ;
	  if( !notReally )
	    FilerMakeDir(pFiler, dosname) ;
	}

	return 0 ;
}


static	int
RestoreFile(FILERCOM *pFiler, const char *name, const struct stat *sbuf)
{
	char	dosname[MAXPATHLEN] ;
	const char	*filename ;
	int	needRestore = 1 ;
	int	foundit = 0 ;

	if( ExcludeFile(name) )
	  return 0 ;

	unix2dos(name, dosname) ;
	filename = lxbasename(dosname) ;

/*
	fprintf(USE_OUT,"dosname is \"%s\" and lxbasename is \"%s\"\n",dosname,filename);
	fprintf(USE_OUT,"RestoreFile was passed \"%s\"\n",name);
*/


	/* Policy:  If user entered "-all" (since = 0), update all files
	 * unconditionally.  If user specified a date (-since or -newer),
	 * update all files newer than the specified date.  Else, find out
	 * the date of the file on the HP (if any) and update if the
	 * local copy is newer.
	 */


	/* first, see if exists, and get its age */

	if( FilerAskDir(pFiler, dosname) != SERVER_ACK )
	  return 1 ;

	while( FilerGetDir(pFiler) == GOT_DIR_ENTRY )
	{
	  if( strmatch((char *)pFiler->Name, filename) )
	  {
/*
	    fprintf(USE_OUT,"foundit\n");
*/
	    foundit = 1 ;
	  }
	}

	if( since != -1 )
	{
	  if( sbuf->st_mtime <= since )
	    needRestore = 0 ;
	}
	else if( foundit && sbuf->st_mtime <= FDate2Unix(&pFiler->DateStamp) )
	  needRestore = 0 ;

	if( needRestore && foundit )
	{
	  if( verbose >= 2 ) fprintf(USE_OUT, "delete %s ... ", dosname) ;
	  if ( notReally ) 
	    fprintf(USE_OUT, "\n") ;
	  else {
	    if( FilerDelFile(pFiler, dosname) != FILE_DELETED )
	    {
	      fprintf(USE_OUT, "could not delete %s\n", dosname) ;
	      needRestore = 0 ;
	    }
	  } /* only delete if not notReally */
	}

	if( needRestore )
	{
	  if( verbose )
	    fprintf(USE_OUT, "restore %s, %ld bytes ... ", name, (long) sbuf->st_size) ;
	  if( notReally )
	    fprintf(USE_OUT, "\n") ;
	  else
	    switch( FilerSendFile(pFiler, (char *)name, dosname) )
	    {
	      case NO_SOURCE_FILE:
		fprintf(USE_OUT, "cannot open %s\n", name) ;
		break ;
	      case CANNOT_SEND_FNAME:
		fprintf(USE_OUT, "cannot send %s\n", dosname) ;
		break ;
	      case NO_RESPONSE:
		fprintf(USE_OUT, "server does not respond\n") ;
		break ;
	      case FILE_SEND_OK:
		if( verbose ) fprintf(USE_OUT, "done\n") ;
		break ;
	    }
	}

	return 0 ;
}



/* build the directory path necessary to house this file */


static	int
mkdir2(char *path, int mode)
{
	char	*last ;

	if( access(path, W_OK|X_OK) == 0 )
	  return 0 ;

	last = strrchr(path,'/') ;
	if( last != NULL ) {
	  *last = '\0' ;
	  mkdir2(path,mode) ;
	  *last = '/' ;
	}
	return mkdir(path,mode) ;
}

static	int
makePath(char *name)
{
	char	path[256], *last ;

	strcpy(path,name) ;

	if( (last = strrchr(path, '/')) == NULL )
	  return 0 ;

	*last = '\0' ;

	if( access(path, W_OK|X_OK) == 0 )
	  return 0 ;

	if( mkdir2(path, 0775) == 0 )
	  return 0 ;

	fprintf(USE_OUT, "lxbackup: unable to make directory %s, %s\n",
		path, sys_errlist[errno]) ;

	return -1 ;
}






	/* convert dos-style name to unix-style name */

static	void
dos2unix(const char *dos, char *unx)
{
	const char	*ptr = dos ;
	char		*ptr2 = unx ;

	while(*ptr != '\0' )
	{
	  if( *ptr == '\\' ) {
	    *ptr2++ = '/' ;
	  }
	  else if ( (noColon == 1) && ( *ptr == ':') ) {
	    *ptr2++ = '-' ;
	  }

	  else
	    *ptr2++ = tolower(*ptr) ;

	  ++ptr ;
	}
	*ptr2 = '\0' ;
}


static	void
colonToDash(const char *unx, char *dos)
{
	const char	*ptr = unx ;
	char		*ptr2 = dos ;
	int		unix2dosCount = 0;

	while(*ptr != '\0' )
	{
	  ++unix2dosCount ;
	  if ( (noColon == 1) && (*ptr == ':') && (unix2dosCount == 2) ) {
	    *ptr2++ = '-' ;
	  }

	  else
	    *ptr2++ = *ptr ;

	  ++ptr ;
	}
	*ptr2 = '\0' ;
}

	/* convert unix-style name to dos-style name */

static	void
unix2dos(const char *unx, char *dos)
{
	const char	*ptr = unx ;
	char		*ptr2 = dos ;
	int		unix2dosCount = 0;

	while(*ptr != '\0' )
	{
	  ++unix2dosCount ;
	  if( *ptr == '/' ) {
	    *ptr2++ = '\\' ;
	  }
	  else if ( (noColon == 1) && (*ptr == '-') && (unix2dosCount == 2) ) {
	    *ptr2++ = ':' ;
	  }

	  else
	    *ptr2++ = toupper(*ptr) ;

	  ++ptr ;
	}
	*ptr2 = '\0' ;
}

	/* examine filename and determine if it needs to be backed up */

static	time_t
FDate2Unix(FDATESTAMP *fdate)
{
	struct tm tm ;

	tm.tm_sec = fdate->sec ;
	tm.tm_min = fdate->min ;
	tm.tm_hour = fdate->hour ;
	tm.tm_mday = fdate->day ;
	tm.tm_mon = fdate->month - 1 ;
	tm.tm_year = fdate->year + 80 ;
	tm.tm_wday = tm.tm_yday = tm.tm_isdst = 0 ;
#ifdef	__STDC__
	return mktime(&tm) ;
#else
	return timelocal(&tm) ;
#endif
}


static	const	char	*
lxbasename(const char *name)
{
	const char	*rval = strrchr(name, '\\') ;
	return rval == NULL ? name : rval+1 ;
}



	/* mmddhhmm[yy] format */

time_t
parseDate(char *date)
{
	time_t		temp ;
	struct tm	tm, *tmptm ;
	int		idate = atoi(date) ;

	temp = time(NULL) ;
	tmptm = localtime(&temp) ;
	tm = *tmptm ;

	if( idate > 99999999 ) {
	  tm.tm_year = idate % 100 ; idate /= 100 ;
	}

	/* if year < 80 (we *are* dealing with DOS, here!)
	   assume it's 21st century -PW */
	if ( tm.tm_year < 80 ) tm.tm_year += 100;

	tm.tm_min = idate % 100 ; idate /= 100 ;
	tm.tm_hour = idate % 100 ; idate /= 100 ;
	tm.tm_mday = idate % 100 ; idate /= 100 ;
	tm.tm_mon = (idate % 100) - 1 ;

#ifdef	__STDC__
	return mktime(&tm) ;
#else
	return timelocal(&tm) ;
#endif
}


time_t
fileTime(char *file)
{
	struct stat	buf ;

	if( stat(file, &buf) ) 
	  return 0 ;
	return buf.st_mtime ;
}


static	void
die(char *msg, char *msg2, int rcode)
{
	if( msg != NULL )
	  fprintf(USE_OUT, "lxbackup: %s\n", msg) ;
	if( msg2 != NULL )
	  fprintf(USE_OUT, "%s\n", msg2) ;
	exit(EXITVAL(rcode)) ;
}
